### PR TITLE
Implemented initial design for the location based access control

### PR DESCRIPTION
**Pull request by suthagar23**
# Initial design for the location based access control project.

## Description

This is an initial design for the location based access control which contains the basic implementation for module. It contains,

1. Common methods for accessing person attributes
2. Common methods for AOP Service
3. Other some Common methods for the project.

## Ticket 

Ticekt : https://issues.openmrs.org/browse/LBAC-1

## Others
- [x] I was able to compile the project without any failures.

### CODE REVIEW

**Comment by suthagar23** [comment]
@dkayiwa  Could you have a look at this PR?

File: api/src/main/java/org/openmrs/module/locationbasedaccess/AOPService.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import org.aopalliance.aop.Advice;
+import org.springframework.aop.Advisor;
+import java.lang.reflect.Method;
+
+/**
+ * It will used to create the abstract methods for using AOP Services inside the classes
+ */
+public interface AOPService extends Advisor {
+
+    boolean matches(Method method, Class targetClass);
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/AOPService.java]: I do not think we should create an interface when we do not yet have a use for it.

File: api/src/main/java/org/openmrs/module/locationbasedaccess/common/UserControlAdvice.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+import org.aopalliance.intercept.MethodInterceptor;
+import org.aopalliance.intercept.MethodInvocation;
+import org.openmrs.Patient;
+import org.openmrs.Person;
+import org.openmrs.User;
+import org.openmrs.PersonAttribute;
+import org.springframework.aop.framework.ReflectiveMethodInvocation;
+import java.util.Set;
+
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/locationbasedaccess/common/UserControlAdvice.java]: What is the use of this class?

**Comment by dkayiwa** [comment]
Let us deal with these one step at a time. Can you start with a real use case from the end user's perspective? Then we implement what is needed to achieve it.

**Comment by dkayiwa** [comment]
For instance, you can start with something as simple as registering a patient in a specific location. Do this in a way that we can even demo it for the end user. That way, we do not end up with classes and constants we may not use.

File: api/src/main/java/org/openmrs/module/locationbasedaccess/common/UserControlAdvice.java (hunk 2 above)
Comment (suthagar23) [reply api/src/main/java/org/openmrs/module/locationbasedaccess/common/UserControlAdvice.java]: It's a common class for the AOP Methods. We can use for users and patients while accessing the service methods.

**Comment by suthagar23** [comment]
Actually, I have tried only for Patients at the beginning, and then I made these common classes which can be used for the users as well. I have created `PatientControlImpl` class which used these common methods to save a location attribute while saving a patient. 
Could you please have a look at here - https://github.com/openmrs/openmrs-module-locationbasedaccess/pull/4/files

Does this make sense for your question?

**Comment by dkayiwa** [comment]
That is too much more than what am asking. Look at it this way. If a custom came and told you that all they need is to simply register patients with locations. That is the only code we need for now.

**Comment by suthagar23** [comment]
Yah I got it well :smile: I will update the PR soon.

**Comment by suthagar23** [comment]
@dkayiwa I moved to [pull/4](https://github.com/openmrs/openmrs-module-locationbasedaccess/pull/4) with the new code base for only patients. Please have a look at there. Thanks.

**Comment by suthagar23** [comment]
I'm closing this PR, and I will open it while starting work on this.

**Pull request closed without merging**